### PR TITLE
EOS-27040 cob: Implement new btree for bytecount table to store pool version information.

### PR DESCRIPTION
**Pull request by kanchan-chaudhari**
Changes:
- Added Bytecount btree inside cob domain.
- Added struct m0_cob_bckey and struct m0_cob_bcrec inside struct m0_cob.
- Added ut "bytecount-ut" to test the bytecount btree code.

Signed-off-by: Kanchan Chaudhari <kanchan.chaudhari@seagate.com>
Signed-off-by: Madhavrao Vemuri <madhav.vemuri@seagate.com>

# Problem Statement
- Implement new btree for bytecount table to store pool version information.

# Design
-  https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/815267859/CORTX+Degraded+byte+count

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: cob/cob.c [hunk 1]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+	if (rc == 0) {
+		/*
+		m0_be_btree_cursor_kv_get(&it->ci_cursor, &key, NULL);
+		bckey = (struct m0_cob_bckey *)key.b_addr;
+		*/
+		/**
+		 * Check if we are stil' inside the object bounds. Assert and
+		 * key copy can be done only in this case.
+		 */
+		/** TODO: Review below condition */
```
Comment (AbhishekSahaSeagate) [inline cob/cob.c:507]: We can remove this commented condition, and its comment above, as we don't need it for bytecount btree.

File: cob/cob.c [hunk 2]
```diff
@@ -521,7 +707,7 @@ M0_INTERNAL int m0_cob_domain_credit_add(struct m0_cob_domain          *dom,
 		return M0_ERR(-ENOMEM);
 	m0_be_0type_add_credit(bedom, &m0_be_cob0, cdid_str, &data, cred);
 	M0_BE_ALLOC_CREDIT_PTR(dom, seg, cred);
-	m0_be_btree_create_credit(&dummy, 5 /* XXX */, cred);
+	m0_be_btree_create_credit(&dummy, 6 /* XXX */, cred);
```
Comment (AbhishekSahaSeagate) [inline cob/cob.c:710]: What is this dummy reference, and hardcoded value '6'.

File: cob/cob.c [hunk 3]
```diff
@@ -877,6 +1070,69 @@ static int cob_ns_lookup(struct m0_cob *cob);
...
+{
+	struct m0_buf key;
+	struct m0_buf val;
+	int           rc;
+
+	M0_PRE(bc_key != NULL &&
+	       m0_fid_is_set(&bc_key->cbk_pfid));
+
+	m0_buf_init(&key, bc_key, m0_cob_bckey_size(bc_key));
+	m0_buf_init(&val, &cob->co_bcrec, sizeof cob->co_bcrec);
```
Comment (AbhishekSahaSeagate) [inline cob/cob.c:1084]: Instead of populating cob->co_bcrec, lets add an out parameter for rec and populate that during lookup.

File: cob/cob.c [hunk 4]
```diff
@@ -1945,12 +2202,16 @@ static void cob_table_tx_credit(struct m0_be_btree *tree,
 			.s_rec = m0_cob_max_nskey_size(),
 				     /* XXX ^^^^^ is it right? */
 		},
+		[COB_KVTYPE_BC] = {
+			.s_key = m0_cob_max_bckey_size(),
```
Comment (AbhishekSahaSeagate) [inline cob/cob.c:2206]: We can change to to call the normal size functions, instead of max size.

File: cob/cob.h [hunk 5]
```diff
@@ -508,6 +509,18 @@ struct m0_cob_earec {
 	char              cer_body[0]; /**< EA body */
 } M0_XCA_RECORD M0_XCA_DOMAIN(be);
 
+/** Byte count table key. */
+struct m0_cob_bckey {
+	struct m0_fid     cbk_pfid;    /**< Pool ver fid */
+	uint64_t          cbk_user_id; /**< User id str? */
```
Comment (AbhishekSahaSeagate) [inline cob/cob.h:515]: remove the 'str?'

File: cob/cob.h [hunk 6]
```diff
@@ -575,6 +588,9 @@ struct m0_cob {
 	struct m0_cob_nsrec    co_nsrec;    /**< object fid, basic stat data */
 	struct m0_cob_fabrec  *co_fabrec;   /**< fileattr_basic data (acl...) */
 	struct m0_cob_omgrec   co_omgrec;   /**< permission data */
+	/* XXX Do we need below structures? XXX */
```
Comment (AbhishekSahaSeagate) [inline cob/cob.h:591]: We can remove these structures below.

**Review by AbhishekSahaSeagate — CHANGES_REQUESTED** [review]

File: cob/cob.c (hunk 2 above)
Comment (kanchan-chaudhari) [reply cob/cob.c:710]: 1. The dummy reference here is set to the segments btree using code `struct m0_be_btree   dummy = { .bb_seg = seg };` and therefore we are allowing credits to create btree.
2. Hardcoded value '6' is the number of credits used when btree's get created in cob domain. Now we have 6 btree's after adding `cd_bytecount` tree in cob domain, so we are using hardcoded value '6'.

File: cob/cob.c [hunk 7]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+M0_INTERNAL int m0_cob_bckey_cmp(const struct m0_cob_bckey *k0,
+				 const struct m0_cob_bckey *k1)
+{
+	int rc;
+
+	M0_PRE(m0_fid_is_set(&k0->cbk_pfid));
+	M0_PRE(m0_fid_is_set(&k1->cbk_pfid));
+
+	rc = m0_fid_cmp(&k0->cbk_pfid, &k1->cbk_pfid);
+	return rc ?: k0->cbk_user_id == k1->cbk_user_id ? 0 : -1;
```
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Done.

File: cob/cob.c [hunk 8]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+/**
+ * Make bytecount key for iterator. Allocate space for maximal possible name.
+ */
+M0_UNUSED static int m0_cob_max_bckey_make(struct m0_cob_bckey **keyh,
+					   const struct m0_fid  *pver_fid,
+					   uint64_t              user_id)
+{
+	struct m0_cob_bckey *key;
+
+	key = m0_alloc(sizeof *key + M0_COB_NAME_MAX);
```
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Corrected.

File: cob/cob.c [hunk 9]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+	rc = m0_fid_cmp(&k0->cbk_pfid, &k1->cbk_pfid);
+	return rc ?: k0->cbk_user_id == k1->cbk_user_id ? 0 : -1;
+}
+
+/**
+ * Maximal possible bckey size.
+ */
+static size_t m0_cob_max_bckey_size(void)
+{
+	return sizeof(struct m0_cob_bckey) + M0_COB_NAME_MAX;
```
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Removed.

File: cob/cob.c [hunk 10]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
 	.ko_compare = ns_cmp
 };
 
+/**
+ * Make bytecount key for iterator. Allocate space for maximal possible name.
+ */
+M0_UNUSED static int m0_cob_max_bckey_make(struct m0_cob_bckey **keyh,
```
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Removed.

File: cob/cob.c [hunk 11]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+					uint64_t                   user_id)
+{
+	int rc;
+
+	/* Prepare entry key using passed started position. */
+	rc = m0_cob_max_bckey_make(&it->ci_key, pver_fid, user_id);
+	if (rc != 0)
+		return M0_RC(rc);
+
+	it->ci_rec = m0_alloc(m0_cob_max_bcrec_size());
```
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Corrected.

File: cob/cob.c [hunk 12]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+
+	m0_buf_init(&key, it->ci_key, m0_cob_bckey_size(it->ci_key));
+	m0_buf_init(&val, it->ci_rec, m0_cob_bcrec_size(it->ci_rec));
+	rc = m0_be_btree_cursor_get_sync(&it->ci_cursor, &key, true);
+	if (rc == 0) {
+		m0_be_btree_cursor_kv_get(&it->ci_cursor, &key, &val);
+		bckey = (struct m0_cob_bckey *)key.b_addr;
+		bcrec = (struct m0_cob_bcrec *)val.b_addr;
+
+		/**
```
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Removed.

File: cob/cob.c [hunk 13]
```diff
@@ -352,6 +352,190 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+{
+	/*
+	struct m0_cob_bckey *bckey;
+	struct m0_buf        key;
+	*/
+	int                  rc;
+
+	rc = m0_be_btree_cursor_next_sync(&it->ci_cursor);
+	if (rc == 0) {
+		/*
```
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Done.

File: cob/cob.c (hunk 1 above)
Comment (kanchan-chaudhari) [reply cob/cob.c]: Done.

File: cob/cob.c (hunk 4 above)
Comment (kanchan-chaudhari) [reply cob/cob.c]: Done.

File: cob/cob.h (hunk 5 above)
Comment (kanchan-chaudhari) [reply cob/cob.h]: Removed.

File: cob/cob.h (hunk 6 above)
Comment (kanchan-chaudhari) [reply cob/cob.h]: Removed.

File: cob/cob.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (kanchan-chaudhari) [reply cob/cob.c]: Fixed.

File: cob/cob.c [hunk 14]
```diff
@@ -352,6 +352,154 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+static m0_bcount_t bc_vsize(const void *val)
+{
+	return sizeof(struct m0_cob_bcrec);
+}
+
+/**
+ * Make bytecount key for iterator. Allocate space for key.
+ */
+static int m0_cob_bckey_make(struct m0_cob_bckey **keyh,
+				       const struct m0_fid  *pver_fid,
```
Comment (madhavemuri) [inline cob/cob.c:385]: Align the variables here.

File: cob/cob.c [hunk 15]
```diff
@@ -352,6 +352,154 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+	struct m0_buf        key;
+	struct m0_buf        val;
+	int                  rc;
+
+	m0_buf_init(&key, it->ci_key, m0_cob_bckey_size());
+	m0_buf_init(&val, it->ci_rec, m0_cob_bcrec_size());
+	rc = m0_be_btree_cursor_get_sync(&it->ci_cursor, &key, true);
+	if (rc == 0) {
+		m0_be_btree_cursor_kv_get(&it->ci_cursor, &key, &val);
+		bckey = (struct m0_cob_bckey *)key.b_addr;
```
Comment (madhavemuri) [inline cob/cob.c:456]: Check if bckey and bcrec can be avoided.

File: cob/cob.c [hunk 16]
```diff
@@ -352,6 +352,154 @@ static const struct m0_be_btree_kv_ops cob_ns_ops = {
...
+
+		M0_ASSERT(sizeof(bckey) <= m0_cob_bckey_size());
+		M0_ASSERT(sizeof(bcrec) <= m0_cob_bcrec_size());
+		memcpy(it->ci_key, bckey, m0_cob_bckey_size());
+		memcpy(it->ci_rec, bcrec, m0_cob_bcrec_size());
+	}
+	return M0_RC(rc);
+}
+
+M0_INTERNAL int m0_cob_bc_iterator_next(struct m0_cob_bc_iterator *it)
```
Comment (madhavemuri) [inline cob/cob.c:467]: [Suggestion] Check if iterator_next and iterator_get can be combined into a single API, with an additional argument to select next operation.

File: cob/cob.c [hunk 17]
```diff
@@ -877,6 +1034,70 @@ static int cob_ns_lookup(struct m0_cob *cob);
...
+	       m0_fid_is_set(&bc_key->cbk_pfid));
+
+	m0_buf_init(&key, bc_key, m0_cob_bckey_size());
+	m0_buf_init(&val, bc_rec, m0_cob_bcrec_size());
+
+	rc = cob_table_lookup(&cob->co_dom->cd_bytecount, &key, &val);
+	if (rc == 0) {
+		cob->co_flags |= M0_CA_BCREC;
+	}
+	return rc;
```
Comment (madhavemuri) [inline cob/cob.c:1055]: Use M0_RC() here

File: cob/cob.c [hunk 18]
```diff
@@ -877,6 +1034,70 @@ static int cob_ns_lookup(struct m0_cob *cob);
...
+	M0_PRE(bc_key != NULL &&
+	       m0_fid_is_set(&bc_key->cbk_pfid));
+
+	m0_buf_init(&key, bc_key, m0_cob_bckey_size());
+	m0_buf_init(&val, bc_rec, m0_cob_bcrec_size());
+
+	rc = cob_table_lookup(&cob->co_dom->cd_bytecount, &key, &val);
+	if (rc == 0) {
+		cob->co_flags |= M0_CA_BCREC;
+	}
```
Comment (madhavemuri) [inline cob/cob.c:1054]: Remove brackets  for single line if statement.

File: cob/ut/bytecount.c [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+{
+	struct m0_be_tx         tx_;
+	struct m0_be_tx        *tx = &tx_;
+	int                     rc;
+	struct m0_be_tx_credit  accum = {};
+	int                     i;
+	struct m0_cob_bckey     dummy_key = {};
+	struct m0_cob_bcrec     out_rec = {};
+
+	ut_tx_open(tx, &accum);
```
Comment (madhavemuri) [inline cob/ut/bytecount.c:171]: Is txn needed for lookup ?

**Review by madhavemuri — APPROVED** [review]

File: cob/ut/bytecount.c (hunk 19 above)
Comment (kanchan-chaudhari) [reply cob/ut/bytecount.c]: No, Removed from the code.

File: cob/cob.c (hunk 18 above)
Comment (kanchan-chaudhari) [reply cob/cob.c]: Done.

File: cob/cob.c (hunk 17 above)
Comment (kanchan-chaudhari) [reply cob/cob.c]: Done

File: cob/cob.c (hunk 14 above)
Comment (kanchan-chaudhari) [reply cob/cob.c]: Done.

File: cob/cob.c (hunk 15 above)
Comment (kanchan-chaudhari) [reply cob/cob.c:456]: Can't avoid, as we are typecasting the memory address to struct here.

**Review by AbhishekSahaSeagate — APPROVED** [review]

File: cob/cob.c (hunk 16 above)
Comment (kanchan-chaudhari) [reply cob/cob.c:467]: m0_cob_bc_iterator_get() operation we expect iterator already initialized and in m0_cob_bc_iterator_next(), we move the cursor position by one. By looking other implementation in file for iterator functions, I think keeping this as it is would be correct fit.

**Pull request merged**